server/sonar-main/src/test/java/org/sonar/application/cluster/HazelcastClusterTest.java

```diff
@@ -284,23 +286,59 @@ public void hazelcast_must_log_through_sl4fj() {
     );
   }
 
-  private class MemoryAppender<E> extends AppenderBase<E> {
-    private final List<E> events = new ArrayList();
+  @Test
+  public void removing_the_last_application_node_must_clear_web_leader() throws InterruptedException {
+    try (AppStateClusterImpl appStateCluster = new AppStateClusterImpl(newSearchSettings())) {
+      TestAppSettings appSettings = newApplicationSettings();
+      appSettings.set(CLUSTER_HOSTS, appStateCluster.getHazelcastCluster().getLocalEndPoint());
+      appSettings.set(CLUSTER_NODE_PORT, "9004");
+      ClusterProperties clusterProperties = new ClusterProperties(appSettings);
+
+      // Simulate a connection from an application node
+      HazelcastCluster appNode = HazelcastCluster.create(clusterProperties);
+      appNode.tryToLockWebLeader();
+      appNode.setOperational(ProcessId.WEB_SERVER);
+      appNode.setOperational(ProcessId.COMPUTE_ENGINE);
+      appNode.registerSonarQubeVersion("6.6.0.22999");
+
+      assertThat(appStateCluster.getLeaderHostName()).isPresent();
+      assertThat(appStateCluster.isOperational(ProcessId.WEB_SERVER, false)).isTrue();
+      assertThat(appStateCluster.isOperational(ProcessId.COMPUTE_ENGINE, false)).isTrue();
+      assertThat(appStateCluster.getHazelcastCluster().getSonarQubeVersion()).isEqualTo("6.6.0.22999");
+
+      // Shutdown the node
+      appNode.close();
+
+      // Propagation of all information take some time, let's wait 5s maximum
+      int counter = 10;
+      while (appStateCluster.getHazelcastCluster().getSonarQubeVersion() != null && counter > 0) {
+        Thread.sleep(500);
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this use of "Thread.sleep()". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2925)